loader/dirent_on_windows.c

```diff
@@ -28,29 +31,30 @@ struct DIR {
     char *name;           /* null-terminated char string */
 };
 
-DIR *opendir(const char *name) {
+DIR *opendir(const struct loader_instance *instance, const char *name) {
     DIR *dir = 0;
 
     if (name && name[0]) {
         size_t base_length = strlen(name);
         const char *all = /* search pattern must end with suitable wildcard */
             strchr("/\\", name[base_length - 1]) ? "*" : "/*";
 
-        if ((dir = (DIR *)loader_instance_tls_heap_alloc(sizeof *dir)) != 0 &&
-            (dir->name = (char *)loader_instance_tls_heap_alloc(base_length + strlen(all) + 1)) != 0) {
+        if ((dir = (DIR *)loader_instance_heap_alloc(instance, sizeof *dir, VK_SYSTEM_ALLOCATION_SCOPE_COMMAND)) != 0 &&
+            (dir->name = (char *)loader_instance_heap_alloc(instance, base_length + strlen(all) + 1,
+                                                            VK_SYSTEM_ALLOCATION_SCOPE_COMMAND)) != 0) {
             strcat(strcpy(dir->name, name), all);
 
             if ((dir->handle = (handle_type)_findfirst(dir->name, &dir->info)) != -1) {
                 dir->result.d_name = 0;
             } else /* rollback */
             {
-                loader_instance_tls_heap_free(dir->name);
-                loader_instance_tls_heap_free(dir);
+                loader_instance_heap_free(instance, dir->name);
+                loader_instance_heap_free(instance, dir);
```

Comment:
What is the difference between these two paths if instance == NULL?

Reply:
loader_instance_heap_free will call `free()` always, instead of possibly using the allocation callbacks if they are set. Generally, the loader calls `malloc()` and `free()` but uses allocation callbacks if set. But in any pre-instance function (where there is no instance, so its always `NULL`), the only thing we can do is use regular `malloc` & `free`.